cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,451 @@
+  }
+
+  /**
+   * Stores an element in the secure store.
+   * @param name Name of the element to store
+   * @param data The data that needs to be securely stored
+   * @param properties Metadata associated with the data
+   * @throws IOException
+   */
+  @Override
+  public void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    SecureStoreMetadata meta = SecureStoreMetadata.of(name, properties);
+    SecureStoreData secureStoreData = new SecureStoreData(meta, data);
+    writeLock.lock();
+    try {
+      keyStore.setKeyEntry(name, new KeyStoreEntry(secureStoreData, meta), password, null);
+      // Attempt to persist the store.
+      flush();
+    } catch (KeyStoreException e) {
+      // We failed to store the key in the key store. Throw an IOException.
+      throw new IOException("Failed to store the key. ", e);
+    } catch (IOException ioe) {
+      // The key was stored in the keystore successfully but we failed to flush it to the file system.
+      // Remove the key from the keystore and throw an IOException.
+      deleteFromStore(keyStore, name, password);
+      throw ioe;
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  private static Key deleteFromStore(KeyStore keyStore, String name, char[] password) throws IOException {
+    Key key = null;
+    try {
+      if (keyStore.containsAlias(name)) {
+        key = keyStore.getKey(name, password);
+        keyStore.deleteEntry(name);
+      }
+    } catch (KeyStoreException | UnrecoverableKeyException | NoSuchAlgorithmException e) {
+      throw new IOException("Failed to delete the key. " + name, e);
+    }
+    return key;
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param name Name of the element to be deleted
+   */
+  @Override
+  public void delete(String name) throws IOException {
+    Key key = null;
+    writeLock.lock();
+    try {
+      key = deleteFromStore(keyStore, name, password);
+      // Attempt to persist the store. If this fails then the keystore file will have an extra key
+      // that the in-memory store does not. This will be remedied the next time a flush happens.
+      // If another flush does not happen and the system is restarted, the only time that file is read,
+      // then we will have an extra key.
+      flush();
+    } catch (IOException ioe) {
+      if (key != null) {
+        try {
+          keyStore.setKeyEntry(name, key, password, null);
+        } catch (KeyStoreException e) {
+          throw new IOException("Failed to put back the key after a flush failure.", e);
+        }
+      }
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A list of {@link SecureStoreMetadata} objects representing the data stored in the store.
+   * @throws IOException
+   */
+  @Override
+  public List<SecureStoreMetadata> list() throws IOException {
+    List<SecureStoreMetadata> list = new ArrayList<>();
```

Comment:
Prefer to move this and the next line inside the try block and have the return statement inside the try block as well. In general it's better to have the declaration closer to where it's being used.